Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,826 @@
+-- For example mnsymbol = "base-4.9.0.0:GHC.Show.C:Show" and the matching
+-- name "base-4.9.0.0:GHC.Show.Show" from the Prelude. The problem seems to be the
+-- module name GHC.Show.C, probably referring to an internal C library.
+--
+-- To get around this, refineExportsItFallbackInternal uses a less strict matching
+-- rule. If the 'stage3' results are empty we fall back to this refiner.
+refineExportsItFallbackInternal :: MySymbol -> [ModuleExports] -> [ModuleExports]
+refineExportsItFallbackInternal mysymbol exports
+    = case separateBy ':' symbol of
+        [p, _, x]   -> map (\e -> e { qualifiedExports = f p x e }) exports
+        _           -> exports
+  where
+    -- Deal with these?
+    symbol = case mysymbol of
+                MySymbolSysQualified  s -> s
+                MySymbolUserQualified s -> s
+
+    -- Check if package name matches and postfix symbol matches (roughly).
+    f p sym export = filter
+                        (\z -> p `isPrefixOf` z && postfixMatch sym z)
+                        (qualifiedExports export)
+
+refineLeadingDot :: MySymbol -> [ModuleExports] -> [ModuleExports]
+refineLeadingDot (MySymbolUserQualified _)           exports = exports
+refineLeadingDot (MySymbolSysQualified symb)         exports = map (\e -> e { qualifiedExports = f leadingDot e }) exports
+  where
+    leadingDot :: String
+    leadingDot = '.' : last (separateBy '.' symb)
+
+    -- f symbol export = filter (symbol ==) thisExports
+    f symbol export = filter (symbol `isSuffixOf`) thisExports
+       where thisExports = qualifiedExports export         -- Things that this module exports.
+
+refineVisibleExports
+    :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => (String -> m (Maybe String))
+    -> [ModuleExports]
+    -> m [ModuleExports]
+refineVisibleExports getHaddockInterfaces exports = mapM f exports
+  where
+    f :: ModuleExports -> m ModuleExports
+    f mexports = do
+        let pname          = mPackageName     mexports -- e.g. "base-4.8.2.0"
+            thisModuleName = mName            mexports -- e.g. "Prelude"
+            qexports       = qualifiedExports mexports -- e.g. ["base-4.8.2.0:GHC.Base.Just", ...]
+        mVisibleExportsMap <- getVisibleExports getHaddockInterfaces pname
+
+        let visibleExportsMap = fromMaybe (throw $ GMEString $ "ImportedFrom: visible exports map is Nothing") mVisibleExportsMap
+        gmLog GmDebug "visibleExportsMap" $ strDoc $ show visibleExportsMap
+
+        let thisModVisibleExports0 = M.lookup thisModuleName visibleExportsMap
+
+        -- On earlier versions of GHC, our qexports list will not be fully qualified, so it will
+        -- look like ["base:GHC.Base.Just", ...] instead of ["base-4.8.2.0:GHC.Base.Just", ...].
+        -- So if thisModVisibleExports0 is Nothing, fall back to searching on a shorter pname.
+        let thisModVisibleExports = case thisModVisibleExports0 of
```

Comment:
This doesn't need an explicit `case`, `thisModVisibleExports0 `mplus` Just pname'` would do.